krb5/krb5
# otp: add support for otp-challenge mechanism

otp-challenge flag requires edata to obtain a challenge message from the
server that must be presented to the user in order to authenticate 
correctly.

In order to do so, otp_edata sends a password-less Access-Request and 
expects Access-Challenge in reply. Then reads the challenge from 
Reply-Message attribute and remembers the State attribute inside a cookie.

Then it reads the cookie in otp_verify flow and appends it to the next 
Access-Request together with the authentication token so the RADIUS server
can match the second message with the initial one.

---

This implements external challenge functionality that is required for OAuth2
authentication via RADIUS protocol and FreeIPA as discussed here:

* http://mailman.mit.edu/pipermail/krbdev/2021-June/013469.html

The functionality is not yet implemented in FreeIPA, therefore it requires a special
FreeIPA patch and a mocked RADIUS server for testing. The Kerberos code is completed
and ready for review, I would like to wait hovewer for FreeIPA side to be finished
before merging. Please, review.

## How to test:

On server:
1. Install https://github.com/pbrezina/freeipa/tree/otp (enables challenge flag for tuser)
2. Clone and run https://github.com/pbrezina/mock-radius on IPA server
3. Create RADIUS proxy config “localhost/127.0.0.1/Secret123” in IPA
4. Create user “tuser” with RADIUS proxy set to localhost in IPA
5. In kadmin.local > getstrs tuser > you should see challenge flag set to true

On client:
```
kinit -n @IPA.VM -c FILE:armor.ccache
kinit -T FILE:armor.ccache tuser@IPA.VM
```

Note: challenge=true is set only for tuser, other RADIUS proxy users will have challenge=false
